service/controller/resource/ipam/azure_machinepool_network_range_getter.go

```diff
@@ -0,0 +1,64 @@
+package ipam
+
+import (
+	"context"
+	"net"
+
+	"github.com/giantswarm/microerror"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/azure-operator/v4/pkg/helpers"
+	"github.com/giantswarm/azure-operator/v4/service/controller/key"
+)
+
+var nodePoolIPMask = net.CIDRMask(24, 32)
+
+type AzureMachinePoolNetworkRangeGetterConfig struct {
+	Client client.Client
+}
+
+// AzureMachinePoolNetworkRangeGetter is a NetworkRangeGetter implementation for node pools.
+type AzureMachinePoolNetworkRangeGetter struct {
+	client client.Client
+}
+
+func NewAzureMachinePoolNetworkRangeGetter(config AzureMachinePoolNetworkRangeGetterConfig) (*AzureMachinePoolNetworkRangeGetter, error) {
+	if config.Client == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Client must not be empty", config)
+	}
+
+	g := &AzureMachinePoolNetworkRangeGetter{
+		client: config.Client,
+	}
+
+	return g, nil
+}
+
+// GetNetworkRange return the tenant cluster virtual network range, because the
+// node pool subnet is getting its IP address range from all available address
+// ranges in the tenant cluster virtual network.
+func (g *AzureMachinePoolNetworkRangeGetter) GetNetworkRange(ctx context.Context, obj interface{}) (net.IPNet, error) {
+	azureMachinePool, err := key.ToAzureMachinePool(obj)
+	if err != nil {
+		return net.IPNet{}, microerror.Mask(err)
+	}
+
+	// Get AzureCluster CR where the NetworkSpec is stored.
+	azureCluster, err := helpers.GetAzureClusterFromMetadata(ctx, g.client, azureMachinePool.ObjectMeta)
+	if err != nil {
+		return net.IPNet{}, microerror.Mask(err)
+	}
+
+	_, ipNet, err := net.ParseCIDR(azureCluster.Spec.NetworkSpec.Vnet.CidrBlock)
```

Comment:
Should we have condition to gracefully handle situation where `azureCluster.Spec.NetworkSpec.Vnet.CidrBlock` is not yet set? This would happen e.g. when creating cluster with default node pool (default behavior in `gsctl`).

Reply:
I'm never fully sure on how to handle those async cases. For example, 5 lines above, the code will fail if the cluster label is not set, but we have controllers that ensure the label is set. Should handle that case more gracefully?

Reply:
Gracefully handling missing (still not set) `Cluster` fields/labels/annotations would be done by returning `nil` from `ipam`'s `EnsureCreated`, right?

Reply:
> I'm never fully sure about how to best handle those async scenarios. For example, 5 lines above, the code will fail if the cluster label is not set, but we have controllers that ensure the label is set. Should handle that case more gracefully?

On principal we should never error on these kind of issues where it will become eventually consistent (after 1-3 reconciliation rounds of different controllers), because otherwise it's counted in monitoring for alerts and that's not what we want. We know that it's ok that initially there are blank fields that might not be ready for full blown cluster creation and we should just defer the execution where possible.

> Gracefully handling missing (still not set) Cluster fields/labels/annotations would be done by returning nil from ipam's EnsureCreated, right?

Yeah :+1:

Reply:
Bailing out done.